### PR TITLE
Rewrite im() in terms of re() and vice versa fixed #10897

### LINKED ISSUE #10897: rewrite im() in terms of re() and vice versa
There is a bug in the rewrite logic in ``complexes.py`` (see https://github.com/sympy/sympy/pull/8359#issuecomment-200560081 and https://github.com/sympy/sympy/pull/8359#issuecomment-200713532). 

The expected output should be:
```py
re(x).rewrite(im) == x - I*im(x)
im(x).rewrite(re) == -I*(x - re(x))
```

### PR DESCRIPTION
**Pull request by nc1337**
Now the ouput is:

```
In [11]: re(x).rewrite(im)
Out[11]: x - I*im(x)

In [12]: im(x).rewrite(re)
Out[12]: -I*(x - re(x))
```

Also added test.

### CODE REVIEW

**Comment by Nitin216** [comment]
Someone review this PR

**Comment by thisch** [comment]
You need to fix the existing tests in https://github.com/sympy/sympy/pull/8359/files

**Comment by Nitin216** [comment]
@thisch  Got that working on it now.

**Comment by smichr** [comment]
makes sense...what was the motivation? Just curious.

**Comment by thisch** [comment]
I had a look at the code in #8359 and discovered this bug. I don't know where it makes sense to rewrite re in terms of im() and vice versa @zanzibar7

File: sympy/functions/elementary/tests/test_complexes.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
     Abs, adjoint, arg, atan2, conjugate, cos, DiracDelta, E, exp, expand,
     Expr, Function, Heaviside, I, im, log, nan, oo, pi, Rational, re, S,
     sign, sin, sqrt, Symbol, symbols, transpose, zoo, exp_polar, Piecewise,
-    Interval, comp, Integral)
+    Interval, comp, Integral,S)
```
Comment (leosartaj) [inline sympy/functions/elementary/tests/test_complexes.py]: You don't need to add ``S``, it is already imported.

**Comment by leosartaj** [comment]
@Nitin216 You also need to change the rewrite functions.

**Comment by zanzibar7** [comment]
These rewrite rules are for simplification of complex algebraic expressions where the real and imaginary parts have been pulled out piecewise.

```
In [1]: from sympy import *
In [2]: from sympy.abc import x
In [3]: re(x) + im(x)
Out[3]: re(x) + im(x)

In [4]: _.simplify()
Out[4]: re(x) + im(x)

In [5]: _.rewrite(re,im)
Out[5]: x
```
Of course, the code above is the exact bug you've identified, and we should be fixing.
`(re(x) + I*im(x)).rewrite(re,im)` should give `x`, as should `(re(x)+I*im(x)).rewrite(im,re)`
Don't remember the exact situation I'd needed it in, but I think it had something to do
with simplifying eigenvalue expressions in terms of trig functions.  Right now,

```
In [1]: from sympy import *

In [2]: from sympy.abc import x

In [3]: I**2
Out[3]: -1

In [4]: y = re(x) + im(x)*I

In [5]: y.simplify()
Out[5]: re(x) + I*im(x)

In [6]: y.rewrite(re,im)
Out[6]: x - im(x) + I*im(x)

In [7]: y.rewrite(im,re)
Out[7]: I*(x - re(x)) + re(x)
```

**Comment by asmeurer** [comment]
@Nitin216 it looks like you overwrite the fix commit with the test commit (I only see tests, not the fix itself, but based on the previous comments the fix was there at some point). 

In general, I would avoid any git commands that require you to force push. Just make changes and commit them (don't use `--amend`), and then push your changes (without `-f`). Don't do any rebasing or `git commit -a`.  If you have merge conflicts you can use `git merge`.

**Comment by Nitin216** [comment]
Sorry Guys for all the inconvenience , I am just correcting it.

**Comment by Nitin216** [comment]
Now someone review this PR again.

File: sympy/core/tests/test_complex.py [hunk 2]
```diff
@@ -210,3 +210,8 @@ def test_issue_5429():
 def test_issue_4124():
     from sympy import oo
     assert expand_complex(I*oo) == oo*I
+
+def test_issue_10897():
```
Comment (thisch) [inline sympy/core/tests/test_complex.py]: you can remove this test since the same code is already in ``test_complexes.py``

Comment (Nitin216) [inline sympy/core/tests/test_complex.py]: got it @thisch

File: sympy/functions/elementary/tests/test_complexes.py [hunk 3]
```diff
@@ -137,8 +137,8 @@ def test_im():
     assert im((1 + sqrt(a + b*I))/2) == \
         (a**2 + b**2)**Rational(1, 4)*sin(atan2(b, a)/2)/2
 
-    assert im(x).rewrite(re) == x - re(x)
-    assert (x + im(y)).rewrite(im, re) == x + y - re(y)
+    assert im(x).rewrite(re) == -S.ImaginaryUnit*(x - re(x))
+    assert (x + im(y)).rewrite(im, re) == x - S.ImaginaryUnit * ( y - re(y))
```
Comment (leosartaj) [inline sympy/functions/elementary/tests/test_complexes.py]: ``x - S.ImaginaryUnit * (y - re(y))``

**Comment by Nitin216** [comment]
If there is no error will someone merge this PR

File: sympy/functions/elementary/tests/test_complexes.py (hunk 3 above)
Comment (Nitin216) [inline sympy/functions/elementary/tests/test_complexes.py]: @leosartaj Is there something incorrect or is it only spaces ??

Comment (leosartaj) [inline sympy/functions/elementary/tests/test_complexes.py]: Just the spaces. If you could please correct them.

Comment (Nitin216) [inline sympy/functions/elementary/tests/test_complexes.py]: @leosartaj Got that.

**Comment by leosartaj** [comment]
+1

**Comment by zanzibar7** [comment]
+1

**Comment by leosartaj** [comment]
Thanks! @Nitin216

**Pull request merged**